src/Shell32.Desktop/Shell32.cs

```diff
@@ -0,0 +1,111 @@
+        /// </remarks>
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderPath.")]
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        public static extern unsafe HResult SHGetFolderPath(
+            IntPtr hwndOwner,
+            CSIDL nFolder,
+            IntPtr hToken,
+            SHGetFolderPathFlags dwFlags,
+            out IntPtr pszPath);
+
+        /// <summary>
+        /// Gets the path of a folder identified by a CSIDL value.
+        /// </summary>
+        /// <param name="rfid">
+        /// A <see cref="Guid"/> value that identifies the folder whose path is to be retrieved.
+        /// </param>
+        /// <param name="dwFlags">Flags that specify special retrieval options. This value can be 0; otherwise, one or more of the <see cref="KNOWN_FOLDER_FLAG"/> values.</param>
+        /// <param name="hToken">
+        /// An access token that represents a particular user. If this parameter is NULL, which is the most common usage, the function requests the known folder for the current user.
+        /// Assigning the hToken parameter a value of -1 indicates the Default User.
+        /// 
+        /// Microsoft Windows 2000 and earlier: Always set this parameter to NULL.
+        /// Windows XP and later: This parameter is usually set to NULL, but you might need to assign a non-NULL value to hToken for those folders that can have multiple users but are treated as belonging to a single user.The most commonly used folder of this type is Documents.
+        /// </param>
+        /// <param name="pszPath">
+        /// A pointer to a null-terminated string of length MAX_PATH which will receive the path.
+        /// If an error occurs or <see cref="HResult.Code.S_FALSE"/> is returned, this string will be empty.
+        /// </param>
+        /// <returns>If this function succeeds, it returns <see cref="HResult.Code.S_OK"/>. Otherwise, it returns an <see cref="HResult"/> error code.</returns>
+        /// <remarks>
+        /// As of Windows Vista, this function is merely a wrapper for <see cref="SHGetKnownFolderPath"/>.
+        /// The returned path does not include a trailing backslash.
+        /// For example, "C:\Users" is returned rather than "C:\Users\".
+        /// </remarks>
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        public static extern unsafe HResult SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            KNOWN_FOLDER_FLAG dwFlags,
+            IntPtr hToken,
+            out IntPtr pszPath);
+
+        [DllImport(nameof(Shell32), SetLastError = true)]
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for SHGetKnownFolderIDList")]
+        public static extern unsafe HResult SHGetFolderLocation(
+            IntPtr hwndOwner,
+            CSIDL nFolder,
+            IntPtr hToken,
+            int dwReserved,
+            out IntPtr pidl);
```

Comment:
Sorry, I tried hard to make it better, but couldn't. I indeed added ITEMIDLIST to that purpose but the out thingy got me nuts, I tried the FriendlyAttribute.Array without success because it complained about being a managed type pointer. I thought of Marshalling int on an Helper function, but had no success also.

Researching on that I found out that most people just use it as IntPtr and pass it along to other funcions instead of trying to Marshall it. The only reference on Marshalling I found was byte[] being "hand" readen and written with auxiliary functions and thought it was not worth publicizing as a blatant hack like this. So I just published the native functions as you see and added a Helper that calls two functions and extracts the string representation of this path instead.

Can you help me on Marshalling it?